cou/cli.py

```diff
@@ -12,140 +12,164 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""Entrypoint to the 'charmed-openstack-upgrader'."""
-import argparse
+"""Entrypoint for 'charmed-openstack-upgrader'."""
 import logging
 import logging.handlers
-import os
-import pathlib
 import sys
-from datetime import datetime
-from typing import Any
+from enum import Enum
+from typing import Optional
 
+from halo import Halo
+
+from cou.commands import parse_args
 from cou.exceptions import COUException
+from cou.logging import setup_logging
+from cou.steps import UpgradeStep
 from cou.steps.analyze import Analysis
 from cou.steps.execute import apply_plan
 from cou.steps.plan import generate_plan
 from cou.utils import juju_utils
 
-COU_DIR_LOG = pathlib.Path(os.getenv("COU_DATA", ""), "log")
 AVAILABLE_OPTIONS = "cas"
 
 logger = logging.getLogger(__name__)
+progress_indicator = Halo(spinner="line", placement="right")
 
 
-def parse_args(args: Any) -> argparse.Namespace:
-    """Parse cli arguments.
+class VerbosityLevel(Enum):
```

Comment:
Instead of this Enum and `get_log_level`, can we create "action" or specific "type" for `--verbose` argument.? Or simple use simple function in `parse_args` function. e.g.
```python
def verbosity(arg: int) -> int:
    match arg:
        case 0:
            return ERROR
        case 1:
            return logging.WARNING
        case 2:
            return logging.INFO
        case 3:
            return logging.DEBUG
        case _:
            raise argparse.ArgumentTypeError(f"{arg!r} is not valid verbosity")
```

Reply:
do you mean by subclassing `argparse.Action`?

Reply:
I'm not quite sure what the advantages are with this approach comparing to using Enum. Could you elaborate more?

Reply:
It can be more integrated with parsing arguments and CLI will failed directly in the parsing phase and not after it + it will raise and proper exception instead of ValueError which is catch by our try except [here](https://github.com/canonical/charmed-openstack-upgrader/blob/main/cou/cli.py).

Reply:
Now I reading next comment and realized that the `ValueError` is only raised if value is not an integer what from CLI call will never happened. I did not realized that.
It's more useful to catch argument error in `parse_args` function and not in `setup_logging(log_level=get_log_level(quiet=args.quiet, verbosity=args.verbosity))` call, but that argument error should never happen., so it's not important. It is only more understandable to be if it's integrated in parsing arguments, phase instead of in later phase of CLI.
Marking as resolved.